src/App.tsx

```diff
@@ -1,26 +1,287 @@
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [deletingIds, setDeletingIds] = useState<number[]>([]);
+
+  const inputRef = useRef<HTMLInputElement>(null);
+
+  const showError = (message: string) => {
+    setErrorMessage(message);
+    setIsErrorHidden(false);
+    setTimeout(() => setIsErrorHidden(true), 3000);
+  };
+
+  useEffect(() => {
+    if (!USER_ID) {
+      return;
+    }
+
+    const loadTodos = async () => {
+      try {
+        const loadedTodos = await getTodos();
+
+        setTodos(loadedTodos);
+      } catch (error) {
+        showError('Unable to load todos');
+      }
+    };
+
+    loadTodos();
+  }, []);
+
+  useEffect(() => {
+    if (inputRef.current && !tempTodo) {
+      inputRef.current.focus();
+    }
+  }, [todos, tempTodo, isErrorHidden]);
+
   if (!USER_ID) {
     return <UserWarning />;
   }
 
+  const handleSubmit = async (e: React.FormEvent) => {
+    e.preventDefault();
+    const trimmedTitle = title.trim();
+
+    if (!trimmedTitle) {
+      showError('Title should not be empty');
+
+      return;
+    }
+
+    const newTodo = {
+      id: 0,
+      userId: USER_ID,
+      title: trimmedTitle,
+      completed: false,
+    };
+
+    setTempTodo(newTodo);
+    setIsErrorHidden(true);
+
+    try {
+      const createdTodo = await addTodo({
+        userId: USER_ID,
+        title: trimmedTitle,
+        completed: false,
+      });
+
+      setTodos(prev => [...prev, createdTodo]);
+      setTitle('');
+    } catch (error) {
+      showError('Unable to add a todo');
+    } finally {
+      setTempTodo(null);
+    }
+  };
+
+  const handleDelete = async (id: number) => {
+    setDeletingIds(prev => [...prev, id]);
+    setIsErrorHidden(true);
+
+    try {
+      await deleteTodo(id);
+      setTodos(prev => prev.filter(todo => todo.id !== id));
+    } catch (error) {
+      showError('Unable to delete a todo');
+    } finally {
+      setDeletingIds(prev => prev.filter(deletingId => deletingId !== id));
+    }
+  };
+
+  const handleClearCompleted = () => {
+    todos.filter(todo => todo.completed).forEach(todo => handleDelete(todo.id));
+  };
+
+  const filteredTodos = todos.filter(todo => {
+    switch (filter) {
+      case 'active':
+        return !todo.completed;
+      case 'completed':
+        return todo.completed;
+      default:
+        return true;
+    }
+  });
+
+  const activeTodosCount = todos.filter(todo => !todo.completed).length;
+  const completedTodosCount = todos.filter(todo => todo.completed).length;
+
   return (
-    <section className="section container">
-      <p className="title is-4">
-        Copy all you need from the prev task:
-        <br />
-        <a href="https://github.com/mate-academy/react_todo-app-loading-todos#react-todo-app-load-todos">
-          React Todo App - Load Todos
-        </a>
-      </p>
-
-      <p className="subtitle">Styles are already copied</p>
-    </section>
+    <div className="todoapp">
```

Comment:
split jsx and logic into smaller components (Header, Footer, TodoList, TodiItem, ErrorNotification etc.)
It is not good practice make such big component, as App.